### DISCUSSION TITLE
Seeed XIAO ESP32S3 giving compilation errors when the code has previously worked

### DISCUSSION THREAD

**Discussion by syedadil19** [pschatzmann/arduino-audio-tools#1059] [Q&A] [answered]
Hi, I've been trying to minimize the size of the microcontroller for my project as much as possible. 

My code:
#include <Arduino.h>
#include "AudioTools.h"
#include "AudioCodecs/CodecADPCM.h"
#include "Sandbox/BLE/AudioBLE.h"

// LED and Microphone Definitions
#define LED_PIN 13 // Change this to the pin where your LED is connected
#define MIC_PIN 34 // Analog pin for the microphone

// Audio Settings
AudioInfo audioInfo(44100, 1, 16); // 44100 Hz, mono, 16-bit
AnalogAudioStream analogMic;        // Analog microphone stream
ADPCMEncoder adpcm(AV_CODEC_ID_ADPCM_IMA_WAV); // ADPCM encoder
AudioBLEServer ble;                 // BLE Audio server
int receivedValue = -1;             // Variable to store the received value

// BLE Hex Print Class for Debugging
class BLEHexPrint : public Print {
public:
    BLEHexPrint(AudioBLEServer &bleServer) : ble(bleServer) {}

    size_t write(uint8_t b) override {
        char hex[3];
        sprintf(hex, "%02X ", b);
        Serial.print(hex);
        return ble.write(&b, 1);
    }

    size_t write(const uint8_t *buffer, size_t size) override {
        size_t n = 0;
        for (size_t i = 0; i < size; i++) {
            n += write(buffer[i]);
        }
        return n;
    }

private:
    AudioBLEServer &ble;
};

BLEHexPrint bleHexPrint(ble);
EncodedAudioStream encoder(&bleHexPrint, &adpcm); // Encoded audio stream
StreamCopy copier(encoder, analogMic);            // Stream copier from mic to BLE

void setup() {
  pinMode(LED_PIN, OUTPUT); // Set the LED pin as output
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Initialize the analog microphone input
  AnalogConfigESP32 cfg = analogMic.defaultConfig(RX_MODE);
  cfg.setInputPin1(MIC_PIN);
  cfg.sample_rate = 44100;
  cfg.bits_per_sample = 16;
  cfg.channels = 1;
  analogMic.begin(cfg);

  // Initialize the ADPCM encoder
  adpcm.begin();

  // Initialize BLE with the name "ble-receive"
  ble.begin("GDP52");

  // Start the encoder with the audio info
  encoder.begin(audioInfo);
}

void loop() {
  // Handle BLE Audio
  if (ble.availableForWrite() > 0) {
    copier.copy();
    ble.updateData();
  }

  // Handle LED Control
  if (ble.available()) {
    receivedValue = ble.read();
    Serial.println(receivedValue);

    if (receivedValue == 0) {
      digitalWrite(LED_PIN, LOW);
    } else if (receivedValue == 1) {
      digitalWrite(LED_PIN, HIGH);
    }
  }
}

The above code could compile with ESP32-PICO-DevkitM-02 but are giving me the following compilation error when used with any ESP32S3 boards including the Seeed XIAO ESP32S3 (Sense) board. I changed LED_PIN to 3 and MIC_PIN to 6 for the XIAO.

C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:12:1: error: 'AnalogAudioStream' does not name a type
 AnalogAudioStream analogMic;        // Analog microphone stream
 ^~~~~~~~~~~~~~~~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:43:28: error: 'analogMic' was not declared in this scope
 StreamCopy copier(encoder, analogMic);            // Stream copier from mic to BLE
                            ^~~~~~~~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:43:28: note: suggested alternative: 'analogWrite'
 StreamCopy copier(encoder, analogMic);            // Stream copier from mic to BLE
                            ^~~~~~~~~
                            analogWrite
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino: In function 'void setup()':
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:51:3: error: 'AnalogConfigESP32' was not declared in this scope
   AnalogConfigESP32 cfg = analogMic.defaultConfig(RX_MODE);
   ^~~~~~~~~~~~~~~~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:52:3: error: 'cfg' was not declared in this scope
   cfg.setInputPin1(MIC_PIN);
   ^~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:56:3: error: 'analogMic' was not declared in this scope
   analogMic.begin(cfg);
   ^~~~~~~~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:56:3: note: suggested alternative: 'analogWrite'
   analogMic.begin(cfg);
   ^~~~~~~~~
   analogWrite

exit status 1

Compilation error: 'AnalogAudioStream' does not name a type

I've tried looking Online but I can't find a definitive reason for it. Would you happen to know the reason?

**Comment by pschatzmann** [comment, marked as answer]
The AnalogAudioStream is using the I2S API of the ESP32. Unfortunatly this is not supported with any of the new processors.
If you upgrade to the new 3.0.0 ESP32 release you can use the implementation which is [based on the new continous API](https://github.com/pschatzmann/arduino-audio-tools/issues/1039) from Espressif

**Reply by syedadil19** [reply]
The update removed the AnalogAudioStream error but it's now giving me this error even with the ESP32-PICO-DevkitM-02:
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino: In function 'void setup()':
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:51:3: error: 'AnalogConfigESP32' was not declared in this scope
   51 |   AnalogConfigESP32 cfg = analogMic.defaultConfig(RX_MODE);
      |   ^~~~~~~~~~~~~~~~~
C:\Users\syeda\Downloads\PWM_generator\PWM_generator\PWM_generator.ino:52:3: error: 'cfg' was not declared in this scope
   52 |   cfg.setInputPin1(MIC_PIN);
      |   ^~~

exit status 1

Compilation error: 'AnalogConfigESP32' was not declared in this scope

I'm guessing this leaves me to use the version 2.0.14 only right?

**Comment by pschatzmann** [comment]
You better use auto instead of giving the class name, since this has changed!
AnalogConfigESP32 is only for the old API!

**Reply by syedadil19** [reply]
I see. I’ll try it out and let you know. Thank you for the guide

**Reply by pschatzmann** [reply]
I should also have a look at the [available parameters](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_analog_config_e_s_p32_v1.html) because this has changed quite a bit driven by the new API

**Reply by syedadil19** [reply]
Alright, please keep me posted. Thank you for your help.

**Reply by syedadil19** [reply]
I changed the AnalogConfigESP32 to auto and that error was solved. But, now there's this error:
C:\Users\mas1u19\Downloads\PWM_generator\PWM_generator.ino: In function 'void setup()':
C:\Users\mas1u19\Downloads\PWM_generator\PWM_generator.ino:52:7: error: 'class audio_tools::AnalogConfigESP32V1' has no member named 'setInputPin1'
   52 |   cfg.setInputPin1(MIC_PIN);
      |       ^~~~~~~~~~~~

exit status 1

Compilation error: 'class audio_tools::AnalogConfigESP32V1' has no member named 'setInputPin1'

I think that since I'm using auto, I do not need the cfg.setInputPin1(MIC_PIN) line anymore right?

**Reply by pschatzmann** [reply]
As I said, look at the available parameters: In order to select the pins you define the ADC channels!
